### PR TITLE
NumPy 2.0 compatiility: (partially) sync with NEP52

### PR DESCRIPTION
**Pull request by ev-br**
This is a rebase/redo of https://github.com/cupy/cupy/pull/8314, which picked way too many conflicts.

Running tests locally under NumPy 2.1.0 gives several failures. Before fixing them, I'd like to clarify the vision:
- CuPy 13.2 aims to support both Numpy 1.26 and 2.x
- in this PR, do we need tests to pass under NumPy 1.x?
- do we keep shims for things removed in NumPy 2.0?
- do we emit DeprecationWarnings for things deprecated in NumPy 2.0?
- do we track NumPy 2.0.0 or can jump straight to 2.1.0? (several fix-ups went to 2.1 IIUC, e.g. `np.copyto`)

### CODE REVIEW

File: tests/cupy_tests/core_tests/test_ndarray_reduction.py [hunk 1]
```diff
@@ -189,70 +189,70 @@ def test_min_inf(self, xp, dtype, exclude_cutensor):
     @testing.numpy_cupy_allclose(contiguous_check=False)
     def test_ptp_all(self, xp, dtype):
         a = testing.shaped_random((2, 3), xp, dtype, order=self.order)
-        return a.ptp()
+        return xp.ptp(a)
```
Comment (asi1024) [inline tests/cupy_tests/core_tests/test_ndarray_reduction.py:192]: The `xp.ptp` tests are already exist in `tests/cupy_tests/statistics_tests/test_order.py`. Could you remove these tests?

File: tests/cupy_tests/creation_tests/test_from_data.py [hunk 2]
```diff
@@ -286,7 +286,7 @@ def test_array_copy_list_of_cupy_with_dtype_char(self, xp, dtype1, dtype2,
     @testing.numpy_cupy_array_equal()
     def test_array_no_copy(self, xp, dtype, order):
         a = testing.shaped_arange((2, 3, 4), xp, dtype)
-        b = xp.array(a, copy=False, order=order)
+        b = xp.array(a, copy=None, order=order)
```
Comment (asi1024) [inline tests/cupy_tests/creation_tests/test_from_data.py:289]: These tests assume that `b` is not copied from `a`. Could you please revert these changes?

**Review by asi1024 — COMMENTED** [review]
@ev-br Thanks for your PR! I am still in the process of reviewing, but I have commented what I can at this point.

File: tests/cupy_tests/core_tests/test_ndarray_reduction.py (hunk 1 above)
Comment (ev-br) [reply tests/cupy_tests/core_tests/test_ndarray_reduction.py]: Removed

File: tests/cupy_tests/creation_tests/test_from_data.py (hunk 2 above)
Comment (ev-br) [reply tests/cupy_tests/creation_tests/test_from_data.py]: This makes tests stricter, but OK, reverted.

**Comment by asi1024** [comment]
While we are still under discussion about the policy which version of NumPy CuPy v14 should support, in parallel we will start merging on support of the NumPy 2.x series with the following policy as a tentative policy. We believe that this policy can be changed with low modification costs, no matter where this discussion ultimately ends up.

CuPy v14
- Behaves compatible with the newest NumPy 2.x series.
- Able to import with NumPy 1.24 or newer.
- CI run for NumPy v1 is with `numpy._set_promotion_state("weak")`.
- CI is also run for NumPy v2.
- Keeps shims for APIs removed in NumPy 2.0.
- Emit DeprecationWarnings for APIs following NumPy 2.0 deprecation.

CuPy v13
- Behaves compatible with NumPy 1.26.
- Able to import with NumPy 1.24 or newer, including NumPy v2.
- CI run for NumPy v1 is without `numpy._set_promotion_state`.
- CI run for NumPy v2 does not check data-type correctness.

File: cupy/_creation/basic.py [hunk 3]
```diff
@@ -369,3 +369,44 @@ def full_like(
     a = cupy.ndarray(shape, dtype, memptr, strides, order)
     cupy.copyto(a, fill_value, casting='unsafe')
     return a
+
+
+# Array API compatible array.astype wrapper
+
+def astype(x, dtype, /, *, copy=True):
```
Comment (asi1024) [inline cupy/_creation/basic.py:376]: Could you add tests for new APIs?

File: tests/cupy_tests/fft_tests/test_fft.py [hunk 4]
```diff
@@ -408,7 +409,11 @@ def test_fft2(self, xp, dtype, order, enable_nd):
         a = testing.shaped_random(self.shape, xp, dtype)
         if order == 'F':
             a = xp.asfortranarray(a)
-        out = xp.fft.fft2(a, s=self.s, axes=self.axes, norm=self.norm)
+
+        with warnings.catch_warnings():
```
Comment (asi1024) [inline tests/cupy_tests/fft_tests/test_fft.py:413]: Could you add `@testing.with_requires("numpy>=2.0")` to the tests that fial in the NumPy 1.x series?

**Comment by asi1024** [comment]
@ev-br Any updates?

File: tests/cupy_tests/fft_tests/test_fft.py (hunk 4 above)
Comment (asi1024) [reply tests/cupy_tests/fft_tests/test_fft.py:372]: Fixed in acd4237 :smiley:

**Comment by ev-br** [comment]
Will do, thanks. ETA is "not today or tomorrow" though.

**Comment by asi1024** [comment]
To aid in pinpointing the cause of the test failure, we intend to divide some of the fixes presented in this PR and create separate PRs for them.

**Comment by EarlMilktea** [comment]
### MEMO (CC: @ev-br )

The changes introduced in 682d5b1f262a8429b3c038b9bf81085360ffac3a seem to be reverted in the latest `numpy` repo.

Please check https://numpy.org/devdocs/reference/generated/numpy.unique.html for details.

**Comment by asi1024** [comment]
All CIs are green!

**Comment by jakirkham** [comment]
Hooray! 🥳 🎊 

Thanks everyone for all of your hard work to get this one in! 🙏 👏
